File: android/app/src/main/kotlin/io/github/sudomegas/tritium/storage/Range.kt
package io.github.sudomegas.tritium.storage

import java.time.YearMonth

/**
 * XTRITIUM §7.2 — the five time ranges the lists filter by, plus a custom
 * pair. Ported from the desktop's `src/shared/range.ts`, function for
 * function.
 *
 * Nothing here reads a clock — `today` always arrives as a parameter, the
 * same discipline [FuelDraft.quickAddDefaults] already uses, so every
 * boundary is testable without freezing time. Dates are compared as
 * `YYYY-MM-DD` strings: fixed-width and zero-padded, so lexical order IS
 * chronological order.
 *
 * This module answers only "what window does a chip mean" and "is this date
 * inside it" — never what gets computed. [Consumption.consumptionById] is
 * always fed a vehicle's whole fuel history; a range here only ever filters
 * rows already rendered, at the screen (AF7.md §1.2).
 */
enum class RangeKey { ALL, YTD, PREVIOUS_YEAR, THIS_MONTH, PREVIOUS_MONTH, CUSTOM }

/** An inclusive window. `null` on either side means open in that direction. */
data class DateBounds(val from: String?, val to: String?)

val ALL_TIME = DateBounds(from = null, to = null)

/** A custom range as the two fields hold it, before parsing. */
data class CustomRange(val from: String?, val to: String?)

private fun pad(value: Int, width: Int): String = value.toString().padStart(width, '0')
private fun firstOfMonth(year: Int, month: Int): String = "${pad(year, 4)}-${pad(month, 2)}-01"
private fun endOfMonth(year: Int, month: Int): String {
    val lastDay = YearMonth.of(year, month).lengthOfMonth()
    return "${pad(year, 4)}-${pad(month, 2)}-${pad(lastDay, 2)}"
}

/**
 * The window a chip means, given today.
 *
 * `to` is today, not the end of the period, for the two ranges still
 * running — a "this month" reaching the 31st would describe days that have
 * not happened (§3.3).
 */
fun boundsFor(key: RangeKey, today: String, custom: CustomRange? = null): DateBounds {
    val year = today.substring(0, 4).toInt()
    val month = today.substring(5, 7).toInt()

    return when (key) {
        RangeKey.ALL -> ALL_TIME
        RangeKey.YTD -> DateBounds(from = firstOfMonth(year, 1), to = today)
        RangeKey.PREVIOUS_YEAR -> DateBounds(from = "${pad(year - 1, 4)}-01-01", to = "${pad(year - 1, 4)}-12-31")
        RangeKey.THIS_MONTH -> DateBounds(from = firstOfMonth(year, month), to = today)
        RangeKey.PREVIOUS_MONTH -> {
            // January reaches back into last December — a naive month - 1
            // produces month zero.
            val y = if (month == 1) year - 1 else year
            val m = if (month == 1) 12 else month - 1
            DateBounds(from = firstOfMonth(y, m), to = endOfMonth(y, m))
        }
        // An unreadable bound is simply not applied — filtering to nothing
        // while the maker is still typing the year would look like data
        // loss (§3.8).
        RangeKey.CUSTOM -> DateBounds(from = custom?.from, to = custom?.to)
    }
}

/**
 * Whether a date falls inside the window. An entry with no date is shown
 * only when the window is open at both ends — it belongs to no period, so
 * any bounded period naming one would claim something not known, but
 * hiding it from "all time" would make it unreachable (§3.8).
 */
fun withinBounds(date: String, bounds: DateBounds): Boolean {
    if (date.isEmpty()) return bounds.from == null && bounds.to == null
    if (bounds.from != null && date < bounds.from) return false
    if (bounds.to != null && date > bounds.to) return false
    return true
}

/** The entries a window admits, in the order they arrived. */
fun <T> filterByBounds(entries: List<T>, bounds: DateBounds, dateOf: (T) -> String): List<T> =
    entries.filter { withinBounds(dateOf(it), bounds) }

File: src/shared/range.ts
// The time ranges of XTRITIUM §7.2, as pure arithmetic over `YYYY-MM-DD`.
//
// §7 gives the lists "the same time-range chips as the charts", so this module
// is written to be used twice: F7's tables now, F8's charts next. It lives in
// shared/ from the start rather than inside a pane a later milestone would have
// to reach into.
//
// Nothing here reads a clock. Today arrives as a string — from `todayIso`, which
// is built from the local calendar and never from UTC (§3.6, and F4's UTC+3
// trap) — so every boundary is testable without freezing time.
//
// Dates are compared as strings. `YYYY-MM-DD` is fixed-width and zero-padded, so
// lexical order IS chronological order, and no Date object is constructed to
// answer a question about a calendar.

/** XTRITIUM §7.2, in the order the chips are shown. */
export const RANGE_KEYS = [
  'all',
  'ytd',
  'previous-year',
  'this-month',
  'previous-month',
  'custom'
] as const

export type RangeKey = (typeof RANGE_KEYS)[number]

export function isRangeKey(value: unknown): value is RangeKey {
  return typeof value === 'string' && (RANGE_KEYS as readonly string[]).includes(value)
}

/** An inclusive window. `null` on either side means open in that direction. */
export interface DateBounds {
  from: string | null
  to: string | null
}

export const ALL_TIME: DateBounds = { from: null, to: null }

/** A custom range as the two fields hold it, before parsing. */
export interface CustomRange {
  from: string | null
  to: string | null
}

function pad(value: number, width: number): string {
  return value.toString().padStart(width, '0')
}

/**
 * The last day of a month, by the real calendar.
 *
 * `Date.UTC(year, month, 0)` is the day before the first of `month`, which is
 * the last day of the month before it — the standard trick, and the reason
 * February is right in a leap year without a table. Only UTC getters are read,
 * so no local offset and no locale enters (§3.6).
 */
function lastDayOf(year: number, month: number): number {
  return new Date(Date.UTC(year, month, 0)).getUTCDate()
}

function firstOfMonth(year: number, month: number): string {
  return `${pad(year, 4)}-${pad(month, 2)}-01`
}

function endOfMonth(year: number, month: number): string {
  return `${pad(year, 4)}-${pad(month, 2)}-${pad(lastDayOf(year, month), 2)}`
}

/**
 * The window a chip means, given today.
 *
 * `to` is today rather than the end of the period for the two ranges that are
 * still running — a "this month" that claimed to reach the 31st would be
 * describing days that have not happened, and §3.3 keeps TRITIUM out of the
 * future in every other respect too.
 */
export function boundsFor(key: RangeKey, today: string, custom?: CustomRange): DateBounds {
  const year = Number.parseInt(today.slice(0, 4), 10)
  const month = Number.parseInt(today.slice(5, 7), 10)

  switch (key) {
    case 'all':
      return ALL_TIME

    case 'ytd':
      return { from: firstOfMonth(year, 1), to: today }

    case 'previous-year':
      return { from: `${pad(year - 1, 4)}-01-01`, to: `${pad(year - 1, 4)}-12-31` }

    case 'this-month':
      return { from: firstOfMonth(year, month), to: today }

    case 'previous-month': {
      // January reaches back into last December, which is the case a naive
      // `month - 1` gets wrong by producing month zero.
      const y = month === 1 ? year - 1 : year
      const m = month === 1 ? 12 : month - 1
      return { from: firstOfMonth(y, m), to: endOfMonth(y, m) }
    }

    case 'custom':
      // An unreadable bound is simply not applied. The alternative — filtering
      // to nothing while the maker is still typing the year — would look like
      // data loss (§3.8: the app does not argue with what is being typed).
      return { from: custom?.from ?? null, to: custom?.to ?? null }
  }
}

/**
 * Whether a date falls inside the window.
 *
 * An entry with no date is shown only when the window is open at both ends. It
 * belongs to no period, so any period that named one would be claiming
 * something about it that is not known — but hiding it from "all time" would
 * make it unreachable, and an unreachable record cannot be repaired (§3.8).
 */
export function withinBounds(date: string, bounds: DateBounds): boolean {
  if (date.length === 0) return bounds.from === null && bounds.to === null
  if (bounds.from !== null && date < bounds.from) return false
  if (bounds.to !== null && date > bounds.to) return false
  return true
}

/** The entries a window admits, in the order they arrived. */
export function filterByBounds<T extends { date: string }>(
  entries: readonly T[],
  bounds: DateBounds
): T[] {
  return entries.filter((entry) => withinBounds(entry.date, bounds))
}
